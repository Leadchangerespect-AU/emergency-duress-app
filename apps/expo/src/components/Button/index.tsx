import React from "react";
import {
  Platform,
  Pressable,
  PressableProps,
  Text,
  TextProps,
} from "react-native";
import { tv, VariantProps } from "tailwind-variants";

import { cn } from "@/utils/cn";

/**
 * Button component styling configuration using tailwind-variants.
 * Supports different colors, sizes, and variants (solid/outline).
 * Sizes range from xs to lg, affecting both text size and padding.
 */
const button = tv({
  slots: {
    pressable:
      "items-center justify-center rounded-md hover:opacity-85 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600",
    text: "",
  },
  variants: {
    color: {
      primary: {
        pressable: "bg-primary",
        text: "text-white",
      },
      secondary: {
        pressable: "bg-secondary",
        text: "text-white",
      },
      destructive: {
        pressable: "bg-red-600",
        text: "text-white",
      },
    },
    size: {
      xs: {
        pressable: "h-8 px-3.5",
        text: "text-xs",
      },
      sm: {
        pressable: "h-9 px-4",
        text: "text-sm",
      },
      md: {
        pressable: "h-10 px-5",
        text: "text-base",
      },
      lg: {
        pressable: "h-11 px-6",
        text: "text-lg",
      },
    },
    variant: {
      solid: {},
      outline: {
        pressable: "border-2 bg-transparent",
      },
    },
    disabled: {
      true: {
        pressable: "cursor-not-allowed bg-gray-500",
        text: "text-gray-400",
      },
    },
  },
  compoundVariants: [
    {
      variant: "outline",
      color: "primary",
      class: {
        pressable: "border-primary",
        text: "text-primary",
      },
    },
    {
      variant: "outline",
      color: "secondary",
      class: {
        pressable: "border-secondary",
        text: "text-secondary",
      },
    },
    {
      variant: "outline",
      color: "destructive",
      class: {
        pressable: "border-red-600",
        text: "text-red-600",
      },
    },
    {
      variant: "outline",
      disabled: true,
      class: {
        pressable: "border-gray-300 bg-transparent",
        text: "text-gray-300",
      },
    },
    {
      variant: "solid",
      class: {
        text: "text-white",
      },
    },
  ],
  defaultVariants: {
    color: "primary",
    size: "md",
    variant: "solid",
  },
});

/**
 * Props for the Button component.
 * @extends {Omit<PressableProps, "disabled" | "children">} - Base props from PressableProps
 * @extends {Pick<TextProps, "children">} - Children prop from TextProps
 * @extends {VariantProps<typeof button>} - Variant props from tailwind-variants
 */
export interface ButtonProps
  extends Omit<PressableProps, "disabled" | "children">,
    Pick<TextProps, "children">,
    VariantProps<typeof button> {
  /** Additional className for text styling */
  textClassName?: string;
}

/**
 * Button component that supports various visual styles and states.
 *
 * @component
 * @example
 * ```tsx
 * // Solid primary button (default)
 * <Button>Click me</Button>
 *
 * // Extra small secondary button
 * <Button size="xs" color="secondary">Small</Button>
 *
 * // Outline variant with secondary color
 * <Button variant="outline" color="secondary">Click me</Button>
 *
 * // Large destructive button
 * <Button color="destructive" size="lg">Delete</Button>
 * ```
 */
const Button = ({
  disabled,
  color,
  size,
  variant,
  children,
  className,
  textClassName,
  ...props
}: ButtonProps) => {
  const slots = button({ disabled, color, size, variant });
  return (
    <Pressable
      className={cn(slots.pressable(), className)}
      disabled={disabled}
      aria-disabled={disabled}
      role="button"
      {...props}
    >
      {Platform.OS === "web" ? (
        <span className={cn(slots.text(), textClassName)}>{children}</span>
      ) : (
        <Text className={cn(slots.text(), textClassName)}>{children}</Text>
      )}
    </Pressable>
  );
};

export default Button;
