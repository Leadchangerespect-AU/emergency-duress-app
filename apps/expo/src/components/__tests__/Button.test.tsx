import React from "react";
import { fireEvent, render, screen } from "@testing-library/react-native";

import Button from "../Button";

describe("Button", () => {
  it("renders with default props", () => {
    render(<Button testID="test-btn">Default Button</Button>);

    expect(screen.getByText("Default Button")).toBeTruthy();
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("applies secondary color variant", () => {
    render(<Button color="secondary">Secondary</Button>);
    expect(screen.getByText("Secondary")).toBeTruthy();
    expect(screen.toJSON()).toMatchSnapshot();
  });

  // Size variant tests
  it("applies extra small size variant", () => {
    render(<Button size="xs">Extra Small</Button>);
    const button = screen.getByRole("button");
    expect(button.props.className).toContain("h-8 px-3.5");
    expect(screen.getByText("Extra Small").props.className).toContain(
      "text-xs",
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("applies small size variant", () => {
    render(<Button size="sm">Small Button</Button>);
    const button = screen.getByRole("button");
    expect(button.props.className).toContain("h-9 px-4");
    expect(screen.getByText("Small Button").props.className).toContain(
      "text-sm",
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("applies medium size variant", () => {
    render(<Button size="md">Medium Button</Button>);
    const button = screen.getByRole("button");
    expect(button.props.className).toContain("h-10 px-5");
    expect(screen.getByText("Medium Button").props.className).toContain(
      "text-base",
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("applies large size variant", () => {
    render(<Button size="lg">Large Button</Button>);
    const button = screen.getByRole("button");
    expect(button.props.className).toContain("h-11 px-6");
    expect(screen.getByText("Large Button").props.className).toContain(
      "text-lg",
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("disables the button", () => {
    render(<Button disabled>Disabled Button</Button>);

    const button = screen.getByRole("button");
    expect(button.props.accessibilityState.disabled).toBe(true);
    expect(screen.getByText("Disabled Button")).toBeTruthy();
  });

  it("does not trigger onPress when disabled", () => {
    const onPress = jest.fn();
    render(
      <Button disabled onPress={onPress}>
        Disabled Button
      </Button>,
    );

    fireEvent.press(screen.getByText("Disabled Button"));
    expect(onPress).not.toHaveBeenCalled();
  });

  it("handles press events", () => {
    const onPress = jest.fn();
    render(<Button onPress={onPress}>Press Me</Button>);

    fireEvent.press(screen.getByText("Press Me"));
    expect(onPress).toHaveBeenCalled();
  });

  it("applies custom className", () => {
    render(<Button className="custom-class">Styled Button</Button>);

    const button = screen.getByRole("button");
    expect(button.props.className).toContain("custom-class");
  });

  it("matches snapshot for different states", () => {
    render(<Button disabled>Disabled Button</Button>);
    expect(screen.toJSON()).toMatchSnapshot();
  });

  // Outline variant tests
  it("renders outline variant with primary color", () => {
    render(
      <Button variant="outline" color="primary">
        Outline Primary
      </Button>,
    );
    const button = screen.getByRole("button");
    expect(button.props.className).toContain("bg-transparent");
    expect(button.props.className).toContain("border");
    expect(button.props.className).toContain("border-primary");
    expect(screen.getByText("Outline Primary").props.className).toContain(
      "text-primary",
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("renders outline variant with secondary color", () => {
    render(
      <Button variant="outline" color="secondary">
        Outline Secondary
      </Button>,
    );
    const button = screen.getByRole("button");
    expect(button.props.className).toContain("bg-transparent");
    expect(button.props.className).toContain("border");
    expect(button.props.className).toContain("border-secondary");
    expect(screen.getByText("Outline Secondary").props.className).toContain(
      "text-secondary",
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("renders outline variant with destructive color", () => {
    render(
      <Button variant="outline" color="destructive">
        Outline Destructive
      </Button>,
    );
    const button = screen.getByRole("button");
    expect(button.props.className).toContain("bg-transparent");
    expect(button.props.className).toContain("border");
    expect(button.props.className).toContain("border-red-600");
    expect(screen.getByText("Outline Destructive").props.className).toContain(
      "text-red-600",
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("applies outline variant with custom styles", () => {
    render(
      <Button
        variant="outline"
        color="primary"
        className="custom-class"
        textClassName="custom-text"
      >
        Custom Outline
      </Button>,
    );
    const button = screen.getByRole("button");
    expect(button.props.className).toContain("custom-class");
    expect(screen.getByText("Custom Outline").props.className).toContain(
      "custom-text",
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
