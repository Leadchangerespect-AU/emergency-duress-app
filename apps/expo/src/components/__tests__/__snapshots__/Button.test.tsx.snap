// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button applies extra small size variant 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  className="items-center justify-center rounded-md hover:opacity-85 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 bg-primary h-8 px-3.5"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="button"
>
  <Text
    className="text-xs text-white"
  >
    Extra Small
  </Text>
</View>
`;

exports[`Button applies large size variant 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  className="items-center justify-center rounded-md hover:opacity-85 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 bg-primary h-11 px-6"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="button"
>
  <Text
    className="text-lg text-white"
  >
    Large Button
  </Text>
</View>
`;

exports[`Button applies medium size variant 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  className="items-center justify-center rounded-md hover:opacity-85 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 bg-primary h-10 px-5"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="button"
>
  <Text
    className="text-base text-white"
  >
    Medium Button
  </Text>
</View>
`;

exports[`Button applies outline variant with custom styles 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  className="items-center justify-center rounded-md hover:opacity-85 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 h-10 px-5 border-2 bg-transparent border-primary custom-class"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="button"
>
  <Text
    className="text-base text-primary custom-text"
  >
    Custom Outline
  </Text>
</View>
`;

exports[`Button applies secondary color variant 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  className="items-center justify-center rounded-md hover:opacity-85 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 bg-secondary h-10 px-5"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="button"
>
  <Text
    className="text-base text-white"
  >
    Secondary
  </Text>
</View>
`;

exports[`Button applies small size variant 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  className="items-center justify-center rounded-md hover:opacity-85 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 bg-primary h-9 px-4"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="button"
>
  <Text
    className="text-sm text-white"
  >
    Small Button
  </Text>
</View>
`;

exports[`Button matches snapshot for different states 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  className="items-center justify-center rounded-md hover:opacity-85 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 h-10 px-5 cursor-not-allowed bg-gray-500"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="button"
>
  <Text
    className="text-base text-white"
  >
    Disabled Button
  </Text>
</View>
`;

exports[`Button renders outline variant with destructive color 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  className="items-center justify-center rounded-md hover:opacity-85 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 h-10 px-5 border-2 bg-transparent border-red-600"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="button"
>
  <Text
    className="text-base text-red-600"
  >
    Outline Destructive
  </Text>
</View>
`;

exports[`Button renders outline variant with primary color 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  className="items-center justify-center rounded-md hover:opacity-85 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 h-10 px-5 border-2 bg-transparent border-primary"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="button"
>
  <Text
    className="text-base text-primary"
  >
    Outline Primary
  </Text>
</View>
`;

exports[`Button renders outline variant with secondary color 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  className="items-center justify-center rounded-md hover:opacity-85 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 h-10 px-5 border-2 bg-transparent border-secondary"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="button"
>
  <Text
    className="text-base text-secondary"
  >
    Outline Secondary
  </Text>
</View>
`;

exports[`Button renders with default props 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  className="items-center justify-center rounded-md hover:opacity-85 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 bg-primary h-10 px-5"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="button"
  testID="test-btn"
>
  <Text
    className="text-base text-white"
  >
    Default Button
  </Text>
</View>
`;
